README.md
# License #

This program is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation, either version 3 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program.  If not, see <http://www.gnu.org/licenses/>.

@copyright 2022 coactum GmbH

# DisCourse #

## Description ##

The DisCourse activity offers the possibility to conduct group discussions in a multi-step procedure.

Based on the concept of the pyramid discussion, the participants of the DisCourse initially write their own text on the discussed topic in a first phase. In the next two phases, they then combine their own positions with those of other participants and thus develop more differentiated texts together, incorporating other points of view. Finally, in the last phase, all DisCourse participants collaborate to develop a final position on the discussed topic based on their previous results.

In this way, the DisCourse enables the participants to work together to develop arguments and perspectives on a topic and to weigh them against and with each other. This allows the participants to develop differentiated positions on a complex topic in a reflexive exchange while making the progress of the discussion visible by displaying the previous preliminary positions.

On the overview page teachers can …

* See all phases and their deadlines and hints specified when the activity was created
* See a brief summary of all DisCourse groups and their submission states and open the group pages for each of these groups
* Switch phases (phases can also be switched automatically by Moodle at the specified deadline)

Students can …

* Also see all phases, their deadlines and the phase hints
* View only their own groups in the different phases and open the group pages for these

On the group page teachers can see …

* The members, the submission state and (if already handed in) the submission of the selected group
* The hint for the phase of the selected group
* The position of each group from the last phase included in this group

Students can additionally hand in a submission for the group if the phase of the group corresponds to the current phase of the DisCourse (the positions of the pre-groups are already pre-filled in the editor).

## Quick installation instructions ##

### Install from git ###
- Navigate to Moodle root folder.
- **git clone https://github.com/coactum/moodle-mod_discourse.git mod/discourse**

### Install from a compressed file ###
- Extract the compressed file data.
- Rename the main folder to discourse.
- Copy to the Moodle mod/ folder.
- Click the 'Notifications' link on the frontpage administration block.

## Dependencies ##
No dependencies.

## Incompatibilities ##
- Incompatible with 3rd party plugin block_sharing_cart (https://moodle.org/plugins/block_sharing_cart). If you try to copy a DisCourse with the Sharing-Cart it may create an incomplete and unusable DisCourse.

CHANGES.md
## Changelog ##
- [1.2.3]:
    - Ensured compatibility with Moodle 4.2.
        - [Layout]: Minor layout fixes because of the new versions of the bootstrap and fontawesome libraries.
        - [Icon]: Added monologo version of the activity icon for current Moodle versions.
        - [Bugfix]: Fix for an error when a discourse group is deleted externally.
    - [Bugfix]: Hide heading and intro in Moodle 4.0 to avoid redundant information.
    - [Chore]: Some code reformating to fix moodle code check errors.

- [1.2.2]:
    - Hotfix to make recent changes for icon and layout compatible with Moodle pre 4.0.

- [1.2.1]:
    - [Bugfix]: Added some margin between current submission and return button on groupview.
    - [Bugfix]: Reduced height of the phase information area on view page if no phase hints are given.
    - [Bugfix]: Default time values for phases now at noon (Phase 1: +7 days noon, Phase 2: +14 days noon, Phase 3: +21 days noon, Phase 4: +28 days noon).
    - [Bugfix]: If a group was deleted now showing "Deleted group" on groupview for the former submissions.
    - [Improvement]: On groupview now showing if group is in active phase or if phase is already closed.
    - [Improvement]: Phase deadlines are now validated when creating or editing a discourse.
    - [Improvement]: Added hint to show teachers if the group mode "Visible groups" is activated so that participants can see all groups and submissions (and not just their own as usual).
    - [Security]: Added sesskey check for manual switching of phases.
    - [Layout]: Heading and intro are not displayed on Moodle 4.1 and above to comply with the new page layout.
    - [Icon]: Activity icon now has the purpose COLLABORATION in Moodle 4.1 and above.

- [1.2.0]:
    - [Bugfix]: Deleting expired contexts via Moodle privacy mechanism now deletes all groups and groupings of a DisCourse (the grouping wich id is stored for the instance in the table mod_discourse and its groups).
    - [Bugfix]: Deleting the whole plugin now removes all groupings connected with DisCourses (and all groups in this groupings).
    - [Bugfix]: Renaming a DisCourse (in the edit settings, not via the quick edit option in the course) now renames groups and groupings connected with the DisCourse to its new name.
    - [Bugfix]: Manual renaming of groups wont break the display of the shortened group names anymore.
    - [Bugfix]: For DisCourse groups messaging is now enabled by default.
    - [Bugfix]: If an invalid groupingid is stored for a DisCourse or the grouping stored for a DisCourse is deleted, other course groups are not displayed in the DisCourse anymore.
    - [Feature]: Added possibility to backup and restore DisCourses via Moodle backup api.
        - [Note]: If you restore a DisCourse within its original course Moodle does not create new groupings and groups. So no grouping is assigned with the new DisCourse and no participants or submissions are recreated.
        - [Note]: Incompatible with 3rd party plugin block_sharing_cart (https://moodle.org/plugins/block_sharing_cart). If you try to copy a DisCourse with the Sharing-Cart it may create an incomplete and unusable DisCourse.
    - [Feature]: DisCourses can now be included in the course reset.
    - [Versions]: Tested for Moodle 3.9, 3.10, 3.11 and 4.0
        - [Note]: Still uses old icon in Moodle 4.0
        - [Known issue]: Minor display issues with some themes in Moodle 4.0 if sidebars are open and display is too small
        - [Known issue]: After duplicating an activity in Moodle 4.0 Moodle returns the wrong activity allowed groups (for the original activity). In this case no groups but an error message is shown and the activity cant be used correctly. Teachers may just wait a few minutes, rename an activity in the course or contact the moodle administrator to clear the moodle cache to fix this issue.


mod_form.php
<?php
// This file is part of Moodle - http://moodle.org/
//
// Moodle is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Moodle is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.

/**
 * The main mod_discourse configuration form.
 *
 * @package     mod_discourse
 * @copyright   2022 coactum GmbH
 * @license     http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
 */

defined('MOODLE_INTERNAL') || die();

require_once($CFG->dirroot.'/course/moodleform_mod.php');
require_once($CFG->dirroot . '/mod/discourse/locallib.php');

/**
 * Module instance settings form.
 *
 * @package    mod_discourse
 * @copyright  2022 coactum GmbH
 * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
 */
class mod_discourse_mod_form extends moodleform_mod {

    /**
     * Defines forms elements
     */
    public function definition() {
        global $CFG, $PAGE;

        $id = optional_param('update', null, PARAM_INT);

        $mform = $this->_form;

        if (isset($id) && $id !== 0) {
            $discourse = discourse::get_discourse_instance($id);

            // For updating group names.
            $mform->addElement('hidden', 'cmid', $id);
            $mform->setType('cmid', PARAM_INT);

            $mform->addElement('hidden', 'oldname', $discourse->get_module_instance()->name);
            if (!empty($CFG->formatstringstriptags)) {
                $mform->setType('oldname', PARAM_TEXT);
            } else {
                $mform->setType('oldname', PARAM_CLEANHTML);
            }
        }

        // Adding the "general" fieldset, where all the common settings are showed.
        $mform->addElement('header', 'general', get_string('general', 'form'));

        // Adding the standard "name" field.
        $mform->addElement('text', 'name', get_string('modulename', 'mod_discourse'), array('size' => '64'));

        if (!empty($CFG->formatstringstriptags)) {
            $mform->setType('name', PARAM_TEXT);
        } else {
            $mform->setType('name', PARAM_CLEANHTML);
        }

        $mform->addRule('name', null, 'required', null, 'client');
        $mform->addRule('name', get_string('maximumchars', '', 255), 'maxlength', 255, 'client');
        $mform->addHelpButton('name', 'modulename', 'mod_discourse');

        // Adding the standard "intro" and "introformat" fields.
        $this->standard_intro_elements();

        // Adding section for phase completion.
        $mform->addElement('header', 'phasecompletion', get_string('phasecompletion', 'mod_discourse'));

        $mform->addElement('date_time_selector', 'deadlinephaseone', get_string('deadlinephaseone', 'mod_discourse'));
        $mform->setDefault('deadlinephaseone', strtotime("+7 days noon"));

        $mform->addElement('date_time_selector', 'deadlinephasetwo', get_string('deadlinephasetwo', 'mod_discourse'));
        $mform->setDefault('deadlinephasetwo', strtotime("+14 days noon"));

        $mform->addElement('date_time_selector', 'deadlinephasethree', get_string('deadlinephasethree', 'mod_discourse'));
        $mform->setDefault('deadlinephasethree', strtotime("+21 days noon"));

        $mform->addElement('date_time_selector', 'deadlinephasefour', get_string('deadlinephasefour', 'mod_discourse'));
        $mform->setDefault('deadlinephasefour', strtotime("+28 days noon"));

        $mform->addElement('advcheckbox', 'autoswitch', get_string('modeautoswitch', 'mod_discourse'),
            get_string('autoswitch', 'mod_discourse'));

        if (isset($id) && $id !== 0 && !isset($discourse->get_module_instance()->autoswitch)) {
            $mform->setDefault('autoswitch', 0);
        } else {
            $mform->setDefault('autoswitch', 1);
        }

        // Adding section for phase hints.
        $mform->addElement('header', 'phaseshints', get_string('phaseshints', 'mod_discourse'));

        $mform->addElement('textarea', 'hintphaseone', get_string('hintphaseone', 'mod_discourse'),
            'wrap="virtual" rows="2" cols="150"');
        $mform->setType('hintphaseone', PARAM_TEXT);

        $mform->addElement('textarea', 'hintphasetwo', get_string('hintphasetwo', 'mod_discourse'),
            'wrap="virtual" rows="2" cols="150"');
        $mform->setType('hintphasetwo', PARAM_TEXT);

        $mform->addElement('textarea', 'hintphasethree', get_string('hintphasethree', 'mod_discourse'),
            'wrap="virtual" rows="2" cols="150"');
        $mform->setType('hintphasethree', PARAM_TEXT);

        $mform->addElement('textarea', 'hintphasefour', get_string('hintphasefour', 'mod_discourse'),
            'wrap="virtual" rows="2" cols="150"');
        $mform->setType('hintphasefour', PARAM_TEXT);

        // Add standard elements.
        $this->standard_coursemodule_elements();

        // Add standard buttons.
        $this->add_action_buttons();
    }


    /**
     * Custom validation for the form.
     * @param array $data Array with all the form data
     * @param array $files Array with files submitted with form
     * @return array $errors Array with errors
     */
    public function validation($data, $files) {
        $errors = array();

        if ($data['deadlinephaseone'] >= $data['deadlinephasetwo']) {
            $errors['deadlinephaseone'] = get_string('errinvalidphasedeadline', 'mod_discourse');
            $errors['deadlinephasetwo'] = get_string('errinvalidphasedeadline', 'mod_discourse');
        }

        if ($data['deadlinephasetwo'] >= $data['deadlinephasethree']) {
            $errors['deadlinephasetwo'] = get_string('errinvalidphasedeadline', 'mod_discourse');
            $errors['deadlinephasethree'] = get_string('errinvalidphasedeadline', 'mod_discourse');
        }

        if ($data['deadlinephasethree'] >= $data['deadlinephasefour']) {
            $errors['deadlinephasethree'] = get_string('errinvalidphasedeadline', 'mod_discourse');
            $errors['deadlinephasefour'] = get_string('errinvalidphasedeadline', 'mod_discourse');
        }

        return $errors;
    }
}
